Cross-compile performance test for aarch64

This is as far as I've gotten with cross-compiling performance_test. 

1) Generate SYSROOT based on ROS instructions (https://github.com/ros2/ros2_documentation/blob/master/source/Tutorials/Cross-compilation.rst#cross-compiling-ros2)
2) Set SYSROOT variable to generated sysroot
3) Set TOOLCHAIN_FILE for aarch64 compiler, linker, etc

(i.e.)
```
set(CMAKE_C_COMPILER /usr/bin/aarch64-linux-gnu-gcc)
set(CMAKE_CXX_COMPILER /usr/bin/aarch64-linux-gnu-g++)
```

4) `mkdir perf_test_ws`
5) cd `perf_test_ws`
6) Clone performance test and osrf_testing_tools
7) `colcon build --event-handlers console_direct+ --cmake-args -DCMAKE_BUILD_TYPE=Release -DCMAKE_TOOLCHAIN_FILE="${TOOLCHAIN_FILE}"`

The sysroot I am using is Ubuntu Bionic w/ `libboost-all-dev` package installed. 

This is the error message I am recieving
```
Starting >>> osrf_testing_tools_cpp
-- Setting C Standard to 11.
-- Setting C++ Standard to 14.
-- Setting SECURITY to ON.
-- The C compiler identification is GNU 5.4.0              
-- The CXX compiler identification is GNU 5.4.0            
-- Check for working C compiler: /usr/bin/aarch64-linux-gnu-gcc
-- Check for working C compiler: /usr/bin/aarch64-linux-gnu-gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done                    
-- Detecting C compile features
-- Detecting C compile features - done                     
-- Check for working CXX compiler: /usr/bin/aarch64-linux-gnu-g++
-- Check for working CXX compiler: /usr/bin/aarch64-linux-gnu-g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done                  
-- Detecting CXX compile features                          
-- Detecting CXX compile features - done                   
-- googletest version '1.8.0' selected, of versions: '1.8.0'
-- building googletest from '/home/naveen.akkapeddi/grand_central/perf_test_ws/src/osrf_testing_tools_cpp/osrf_testing_tools_cpp/vendor/google/googletest/release-1.8.0.tar.gz'...
-- Configuring done
-- Generating done
-- Build files have been written to: /home/naveen.akkapeddi/grand_central/perf_test_ws/build/osrf_testing_tools_cpp/googletest-1.8.0-extracted
Scanning dependencies of target googletest                 
[ 11%] Creating directories for 'googletest'
[ 22%] Performing download step (verify and extract) for 'googletest'
-- verifying file...
     file='/home/naveen.akkapeddi/grand_central/perf_test_ws/src/osrf_testing_tools_cpp/osrf_testing_tools_cpp/vendor/google/googletest/release-1.8.0.tar.gz'
-- verifying file... done
-- extracting...
     src='/home/naveen.akkapeddi/grand_central/perf_test_ws/src/osrf_testing_tools_cpp/osrf_testing_tools_cpp/vendor/google/googletest/release-1.8.0.tar.gz'
     dst='/home/naveen.akkapeddi/grand_central/perf_test_ws/build/osrf_testing_tools_cpp/googletest-1.8.0-extracted/googletest-1.8.0-src'
-- extracting... [tar xfz]
-- extracting... [analysis]                                    
-- extracting... [rename]
-- extracting... [clean up]
-- extracting... done
[ 33%] No patch step for 'googletest'
[ 44%] No update step for 'googletest'
[ 55%] No configure step for 'googletest'                      
[ 66%] No build step for 'googletest'
[ 77%] No install step for 'googletest'
[ 88%] No test step for 'googletest'                           
[100%] Completed 'googletest'
[100%] Built target googletest
-- Found PythonInterp: /usr/bin/python (found version "2.7.12") 
-- Looking for pthread.h
-- Looking for pthread.h - found                                
-- Looking for pthread_create
-- Looking for pthread_create - not found                       
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found           
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found                
-- Found Threads: TRUE  
-- Configuring done
-- Generating done                                              
-- Build files have been written to: /home/naveen.akkapeddi/grand_central/perf_test_ws/build/osrf_testing_tools_cpp
Scanning dependencies of target gtest                      
Scanning dependencies of target test_runner
Scanning dependencies of target memory_tools
[  4%] Building CXX object src/test_runner/CMakeFiles/test_runner.dir/main.cpp.o
[  8%] Building CXX object googletest-1.8.0-extracted/googletest-1.8.0-build/googlemock/gtest/CMakeFiles/gtest.dir/src/gtest-all.cc.o
[ 12%] Building CXX object src/memory_tools/CMakeFiles/memory_tools.dir/implementation_monitoring_override.cpp.o
[ 16%] Building CXX object src/memory_tools/CMakeFiles/memory_tools.dir/memory_tools_service.cpp.o
[ 20%] Building CXX object src/memory_tools/CMakeFiles/memory_tools.dir/is_working.cpp.o
[ 24%] Building CXX object src/memory_tools/CMakeFiles/memory_tools.dir/testing_helpers.cpp.o
[ 28%] Building CXX object src/memory_tools/CMakeFiles/memory_tools.dir/custom_memory_functions.cpp.o
[ 32%] Building CXX object src/memory_tools/CMakeFiles/memory_tools.dir/monitoring.cpp.o
[ 36%] Building CXX object src/memory_tools/CMakeFiles/memory_tools.dir/initialize.cpp.o
[ 40%] Building CXX object src/memory_tools/CMakeFiles/memory_tools.dir/stack_trace.cpp.o
[ 44%] Building CXX object src/memory_tools/CMakeFiles/memory_tools.dir/verbosity.cpp.o
[ 48%] Building CXX object src/memory_tools/CMakeFiles/memory_tools.dir/register_hooks.cpp.o
[ 52%] Linking CXX executable test_runner                      
[ 52%] Built target test_runner                                
[ 56%] Linking CXX shared library libmemory_tools.so           
[ 56%] Built target memory_tools                               
Scanning dependencies of target memory_tools_interpose
[ 60%] Building CXX object src/memory_tools/CMakeFiles/memory_tools_interpose.dir/memory_tools.cpp.o
[ 64%] Linking CXX shared library libmemory_tools_interpose.so 
[ 64%] Built target memory_tools_interpose
[ 68%] Linking CXX static library libgtest.a                   
[ 68%] Built target gtest
Scanning dependencies of target gtest_main                     
[ 72%] Building CXX object googletest-1.8.0-extracted/googletest-1.8.0-build/googlemock/gtest/CMakeFiles/gtest_main.dir/src/gtest_main.cc.o
[ 76%] Linking CXX static library libgtest_main.a              
[ 76%] Built target gtest_main
Scanning dependencies of target test_memory_tools
Scanning dependencies of target assert_env_vars
Scanning dependencies of target test_parse_environment_variable
[ 80%] Building CXX object test/memory_tools/CMakeFiles/test_memory_tools.dir/test_memory_tools.cpp.o
[ 84%] Building CXX object test/test_runner/CMakeFiles/test_parse_environment_variable.dir/test_parse_environment_variable.cpp.o
[ 88%] Building CXX object test/test_runner/CMakeFiles/assert_env_vars.dir/assert_env_vars.cpp.o
[ 92%] Linking CXX executable test_parse_environment_variable    
[ 92%] Built target test_parse_environment_variable              
[ 96%] Linking CXX executable assert_env_vars                    
[ 96%] Built target assert_env_vars
[100%] Linking CXX executable test_memory_tools                  
/home/naveen.akkapeddi/grand_central/sysroot_arm64v8_ubuntu_bionic_libssl_boost/usr/lib/aarch64-linux-gnu/libdl.so: undefined reference to `_dl_catch_error@GLIBC_PRIVATE'
/home/naveen.akkapeddi/grand_central/sysroot_arm64v8_ubuntu_bionic_libssl_boost/usr/lib/aarch64-linux-gnu/libdl.so: undefined reference to `_dl_signal_error@GLIBC_PRIVATE'
collect2: error: ld returned 1 exit status
make[2]: *** [test/memory_tools/test_memory_tools] Error 1
test/memory_tools/CMakeFiles/test_memory_tools.dir/build.make:98: recipe for target 'test/memory_tools/test_memory_tools' failed
CMakeFiles/Makefile2:1383: recipe for target 'test/memory_tools/CMakeFiles/test_memory_tools.dir/all' failed
make[1]: *** [test/memory_tools/CMakeFiles/test_memory_tools.dir/all] Error 2
make: *** [all] Error 2
Makefile:138: recipe for target 'all' failed
--- stderr: osrf_testing_tools_cpp
/home/naveen.akkapeddi/grand_central/sysroot_arm64v8_ubuntu_bionic_libssl_boost/usr/lib/aarch64-linux-gnu/libdl.so: undefined reference to `_dl_catch_error@GLIBC_PRIVATE'
/home/naveen.akkapeddi/grand_central/sysroot_arm64v8_ubuntu_bionic_libssl_boost/usr/lib/aarch64-linux-gnu/libdl.so: undefined reference to `_dl_signal_error@GLIBC_PRIVATE'
collect2: error: ld returned 1 exit status
make[2]: *** [test/memory_tools/test_memory_tools] Error 1
make[1]: *** [test/memory_tools/CMakeFiles/test_memory_tools.dir/all] Error 2
make: *** [all] Error 2
---
Failed   <<< osrf_testing_tools_cpp	[ Exited with code 2 ]

Summary: 0 packages finished [11.9s]
  1 package failed: osrf_testing_tools_cpp
  1 package had stderr output: osrf_testing_tools_cpp
  2 packages not processed

```

@esteve

--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -13,7 +13,7 @@ jobs:
         sudo apt-get update && sudo apt-get install -y --no-install-recommends qemu-user-static
       displayName: Install QEMU static emulator
     - script: |
-        docker build --build-arg AUTOWARE_DOCKER_ARCH=arm64v8 --build-arg AUTOWARE_TARGET_ARCH=aarch64 -t performance_test_cross -f tools/Dockerimage.crossbuild .
+        docker build --build-arg PERFORMANCE_TEST_DOCKER_ARCH=arm64v8 --build-arg PERFORMANCE_TEST_TARGET_ARCH=aarch64 -t performance_test_cross -f tools/Dockerimage.crossbuild .
       displayName: Build Docker image
     - script: |
         docker run -t -v `pwd`:`pwd` -w `pwd` \

--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -7,6 +7,14 @@ trigger:
     - README.md
 
 jobs:
+- job: build_cross_image
+  steps:
+    -script: |
+      docker build --build-arg AUTOWARE_DOCKER_ARCH=arm64v8 --build-arg AUTOWARE_TARGET_ARCH=aarch64 -t performance_test_cross -f tools/Dockerimage.crossbuild .
+      docker run -it -v ${PWD}:${PWD} -w ${PWD} \
+        --rm performance_test_cross \
+        sh -c "colcon build --merge-install --cmake-args -DCMAKE_TOOLCHAIN_FILE=`pwd`/tools/toolchain.cmake -DBUILD_TESTING=OFF"
+
 - job: ubuntu_bionic
   container:
     image: ros:dashing

--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -6,49 +6,50 @@ trigger:
     exclude:
     - README.md
 
-container:
-  image: ros:dashing
-  options: "--name ci-container -v /usr/bin/docker:/tmp/docker:ro"
-
-steps:
-  - script: |
-      /tmp/docker exec -t -u 0 ci-container \
-      sh -c "apt-get update && DEBIAN_FRONTEND=noninteractive apt-get -o Dpkg::Options::="--force-confold" -y install sudo"
-    displayName: Set up sudo
-  - script: |
-      sudo apt-get install -y maven default-jdk
-    displayName: Install Cyclone DDS dependencies
-  - script: |
-      cd ${PIPELINE_WORKSPACE}
-      git clone https://github.com/eclipse-cyclonedds/cyclonedds.git
-      cd cyclonedds
-      git checkout 4e805597631ed0dcbdc0eecfe9d532cb75180ae7 # Head as of 20190725
-      mkdir -p build
-      cd build
-      cmake -DCMAKE_INSTALL_PREFIX=/opt/cyclonedds ../src
-      cmake --build .
-      sudo cmake --build . --target install
-      find /opt/cyclonedds
-    displayName: Build and install Cyclone DDS
-  - script: sudo apt-get install -y ros-dashing-osrf-testing-tools-cpp default-jre
-    displayName: Install APT dependencies
-  - script: sudo apt-get install -y python3-colcon-common-extensions
-    displayName: Install colcon
-  - script: rosdep update && rosdep install -y --from performance_test --ignore-src
-    displayName: Install ROS dependencies
-  - script: |
-      source /opt/ros/dashing/setup.bash
-      colcon build \
-        --merge-install \
-        --cmake-args \
-          -DCMAKE_BUILD_TYPE=Release \
-          -DPERFORMANCE_TEST_FORCE_CYCLONEDDS=BOOL:ON \
-        --base-path performance_test
-    displayName: Build performance_test
-  - script: |
-      source /opt/ros/dashing/setup.bash
-      colcon test \
-        --merge-install \
-        --base-path performance_test
-      colcon test-result --verbose
-    displayName: Test performance_test
+jobs:
+- job: ubuntu_bionic
+  container:
+    image: ros:dashing
+    options: "--name ci-container -v /usr/bin/docker:/tmp/docker:ro"
+  steps:
+    - script: |
+        /tmp/docker exec -t -u 0 ci-container \
+        sh -c "apt-get update && DEBIAN_FRONTEND=noninteractive apt-get -o Dpkg::Options::="--force-confold" -y install sudo"
+      displayName: Set up sudo
+    - script: |
+        sudo apt-get install -y maven default-jdk
+      displayName: Install Cyclone DDS dependencies
+    - script: |
+        cd ${PIPELINE_WORKSPACE}
+        git clone https://github.com/eclipse-cyclonedds/cyclonedds.git
+        cd cyclonedds
+        git checkout 4e805597631ed0dcbdc0eecfe9d532cb75180ae7 # Head as of 20190725
+        mkdir -p build
+        cd build
+        cmake -DCMAKE_INSTALL_PREFIX=/opt/cyclonedds ../src
+        cmake --build .
+        sudo cmake --build . --target install
+        find /opt/cyclonedds
+      displayName: Build and install Cyclone DDS
+    - script: sudo apt-get install -y ros-dashing-osrf-testing-tools-cpp default-jre
+      displayName: Install APT dependencies
+    - script: sudo apt-get install -y python3-colcon-common-extensions
+      displayName: Install colcon
+    - script: rosdep update && rosdep install -y --from performance_test --ignore-src
+      displayName: Install ROS dependencies
+    - script: |
+        source /opt/ros/dashing/setup.bash
+        colcon build \
+          --merge-install \
+          --cmake-args \
+            -DCMAKE_BUILD_TYPE=Release \
+            -DPERFORMANCE_TEST_FORCE_CYCLONEDDS=BOOL:ON \
+          --base-path performance_test
+      displayName: Build performance_test
+    - script: |
+        source /opt/ros/dashing/setup.bash
+        colcon test \
+          --merge-install \
+          --base-path performance_test
+        colcon test-result --verbose
+      displayName: Test performance_test

--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -9,7 +9,7 @@ trigger:
 jobs:
 - job: build_cross_image
   steps:
-    -script: |
+    - script: |
       docker build --build-arg AUTOWARE_DOCKER_ARCH=arm64v8 --build-arg AUTOWARE_TARGET_ARCH=aarch64 -t performance_test_cross -f tools/Dockerimage.crossbuild .
       docker run -it -v `pwd`:`pwd` -w `pwd` \
         --rm performance_test_cross \

--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -16,7 +16,7 @@ jobs:
         docker build --build-arg AUTOWARE_DOCKER_ARCH=arm64v8 --build-arg AUTOWARE_TARGET_ARCH=aarch64 -t performance_test_cross -f tools/Dockerimage.crossbuild .
       displayName: Build Docker image
     - script: |
-        docker run -it -v `pwd`:`pwd` -w `pwd` \
+        docker run -t -v `pwd`:`pwd` -w `pwd` \
           --rm performance_test_cross \
           sh -c "colcon build --merge-install --cmake-args -DCMAKE_TOOLCHAIN_FILE=`pwd`/tools/toolchain.cmake -DBUILD_TESTING=OFF"
       displayName: Build performance_test

--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -10,10 +10,13 @@ jobs:
 - job: build_cross_image
   steps:
     - script: |
-      docker build --build-arg AUTOWARE_DOCKER_ARCH=arm64v8 --build-arg AUTOWARE_TARGET_ARCH=aarch64 -t performance_test_cross -f tools/Dockerimage.crossbuild .
-      docker run -it -v `pwd`:`pwd` -w `pwd` \
-        --rm performance_test_cross \
-        sh -c "colcon build --merge-install --cmake-args -DCMAKE_TOOLCHAIN_FILE=`pwd`/tools/toolchain.cmake -DBUILD_TESTING=OFF"
+        docker build --build-arg AUTOWARE_DOCKER_ARCH=arm64v8 --build-arg AUTOWARE_TARGET_ARCH=aarch64 -t performance_test_cross -f tools/Dockerimage.crossbuild .
+      displayName: Build Docker image
+    - script: |
+        docker run -it -v `pwd`:`pwd` -w `pwd` \
+          --rm performance_test_cross \
+          sh -c "colcon build --merge-install --cmake-args -DCMAKE_TOOLCHAIN_FILE=`pwd`/tools/toolchain.cmake -DBUILD_TESTING=OFF"
+      displayName: Build performance_test
 
 - job: ubuntu_bionic
   container:

--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -10,6 +10,9 @@ jobs:
 - job: build_cross_image
   steps:
     - script: |
+        apt-get update && apt-get install -y --no-install-recommends qemu-user-static
+      displayName: Install QEMU static emulator
+    - script: |
         docker build --build-arg AUTOWARE_DOCKER_ARCH=arm64v8 --build-arg AUTOWARE_TARGET_ARCH=aarch64 -t performance_test_cross -f tools/Dockerimage.crossbuild .
       displayName: Build Docker image
     - script: |

--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -10,7 +10,7 @@ jobs:
 - job: build_cross_image
   steps:
     - script: |
-        apt-get update && apt-get install -y --no-install-recommends qemu-user-static
+        sudo apt-get update && sudo apt-get install -y --no-install-recommends qemu-user-static
       displayName: Install QEMU static emulator
     - script: |
         docker build --build-arg AUTOWARE_DOCKER_ARCH=arm64v8 --build-arg AUTOWARE_TARGET_ARCH=aarch64 -t performance_test_cross -f tools/Dockerimage.crossbuild .

--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -11,7 +11,7 @@ jobs:
   steps:
     -script: |
       docker build --build-arg AUTOWARE_DOCKER_ARCH=arm64v8 --build-arg AUTOWARE_TARGET_ARCH=aarch64 -t performance_test_cross -f tools/Dockerimage.crossbuild .
-      docker run -it -v ${PWD}:${PWD} -w ${PWD} \
+      docker run -it -v `pwd`:`pwd` -w `pwd` \
         --rm performance_test_cross \
         sh -c "colcon build --merge-install --cmake-args -DCMAKE_TOOLCHAIN_FILE=`pwd`/tools/toolchain.cmake -DBUILD_TESTING=OFF"
 

--- a/tools/Dockerimage.crossbuild
+++ b/tools/Dockerimage.crossbuild
@@ -16,6 +16,7 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
   ros-dashing-ament-cmake-nose \
   ros-dashing-launch-testing
 
+# Convert all absolute symbolic links to relative so that they can be found when copied to the sysroot
 # hadolint ignore=SC2156
 RUN find / -depth -xdev -type l -lname '/*' -exec sh -c 'linkpath="$(readlink {})" && rm -f {} && ln -frsv "$linkpath" "{}"' \;
 

--- a/tools/Dockerimage.crossbuild
+++ b/tools/Dockerimage.crossbuild
@@ -1,13 +1,13 @@
-ARG AUTOWARE_DOCKER_ARCH
-ARG AUTOWARE_TARGET_ARCH
-FROM multiarch/alpine:${AUTOWARE_TARGET_ARCH}-latest-stable AS bootstrap
+ARG PERFORMANCE_TEST_DOCKER_ARCH
+ARG PERFORMANCE_TEST_TARGET_ARCH
+FROM multiarch/alpine:${PERFORMANCE_TEST_TARGET_ARCH}-latest-stable AS bootstrap
 
-FROM ${AUTOWARE_DOCKER_ARCH}/ros:dashing AS sysroot
-ARG AUTOWARE_TARGET_ARCH
+FROM ${PERFORMANCE_TEST_DOCKER_ARCH}/ros:dashing AS sysroot
+ARG PERFORMANCE_TEST_TARGET_ARCH
 
 SHELL ["/bin/bash", "-o", "pipefail", "-c"]
 
-COPY --from=bootstrap /usr/bin/qemu-${AUTOWARE_TARGET_ARCH}-static /usr/bin/qemu-${AUTOWARE_TARGET_ARCH}-static
+COPY --from=bootstrap /usr/bin/qemu-${PERFORMANCE_TEST_TARGET_ARCH}-static /usr/bin/qemu-${PERFORMANCE_TEST_TARGET_ARCH}-static
 
 RUN apt-get update && apt-get install -y --no-install-recommends \
   libboost-all-dev \
@@ -20,18 +20,18 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
 RUN find / -depth -xdev -type l -lname '/*' -exec sh -c 'linkpath="$(readlink {})" && rm -f {} && ln -frsv "$linkpath" "{}"' \;
 
 FROM ros:dashing AS builder
-ARG AUTOWARE_TARGET_ARCH
-ARG AUTOWARE_TARGET_PLATFORM
+ARG PERFORMANCE_TEST_TARGET_ARCH
+ARG PERFORMANCE_TEST_TARGET_PLATFORM
 
 SHELL ["/bin/bash", "-o", "pipefail", "-c"]
 
-ENV AUTOWARE_SYSROOT /sysroot/${AUTOWARE_TARGET_PLATFORM}
-COPY --from=sysroot /lib ${AUTOWARE_SYSROOT}/lib
-COPY --from=sysroot /usr/include ${AUTOWARE_SYSROOT}/usr/include
-COPY --from=sysroot /usr/lib ${AUTOWARE_SYSROOT}/usr/lib
-COPY --from=sysroot /usr/share/pkgconfig ${AUTOWARE_SYSROOT}/usr/share/pkgconfig
-COPY --from=sysroot /opt ${AUTOWARE_SYSROOT}/opt
-COPY --from=sysroot /etc/alternatives ${AUTOWARE_SYSROOT}/etc/alternatives
+ENV PERFORMANCE_TEST_SYSROOT /sysroot/${PERFORMANCE_TEST_TARGET_PLATFORM}
+COPY --from=sysroot /lib ${PERFORMANCE_TEST_SYSROOT}/lib
+COPY --from=sysroot /usr/include ${PERFORMANCE_TEST_SYSROOT}/usr/include
+COPY --from=sysroot /usr/lib ${PERFORMANCE_TEST_SYSROOT}/usr/lib
+COPY --from=sysroot /usr/share/pkgconfig ${PERFORMANCE_TEST_SYSROOT}/usr/share/pkgconfig
+COPY --from=sysroot /opt ${PERFORMANCE_TEST_SYSROOT}/opt
+COPY --from=sysroot /etc/alternatives ${PERFORMANCE_TEST_SYSROOT}/etc/alternatives
 
 # Fix Bionic-specific issues
 # ignore pinning package versions for apt

--- a/tools/Dockerimage.crossbuild
+++ b/tools/Dockerimage.crossbuild
@@ -1,0 +1,47 @@
+ARG AUTOWARE_DOCKER_ARCH
+ARG AUTOWARE_TARGET_ARCH
+FROM multiarch/alpine:${AUTOWARE_TARGET_ARCH}-latest-stable AS bootstrap
+
+FROM ${AUTOWARE_DOCKER_ARCH}/ros:dashing AS sysroot
+ARG AUTOWARE_TARGET_ARCH
+
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+
+COPY --from=bootstrap /usr/bin/qemu-${AUTOWARE_TARGET_ARCH}-static /usr/bin/qemu-${AUTOWARE_TARGET_ARCH}-static
+
+RUN apt-get update && apt-get install -y --no-install-recommends \
+  libboost-all-dev \
+  libtinyxml2-dev \
+  ros-dashing-osrf-testing-tools-cpp \
+  ros-dashing-ament-cmake-nose \
+  ros-dashing-launch-testing
+
+# hadolint ignore=SC2156
+RUN find / -depth -xdev -type l -lname '/*' -exec sh -c 'linkpath="$(readlink {})" && rm -f {} && ln -frsv "$linkpath" "{}"' \;
+
+FROM ros:dashing AS builder
+ARG AUTOWARE_TARGET_ARCH
+ARG AUTOWARE_TARGET_PLATFORM
+
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+
+ENV AUTOWARE_SYSROOT /sysroot/${AUTOWARE_TARGET_PLATFORM}
+COPY --from=sysroot /lib ${AUTOWARE_SYSROOT}/lib
+COPY --from=sysroot /usr/include ${AUTOWARE_SYSROOT}/usr/include
+COPY --from=sysroot /usr/lib ${AUTOWARE_SYSROOT}/usr/lib
+COPY --from=sysroot /usr/share/pkgconfig ${AUTOWARE_SYSROOT}/usr/share/pkgconfig
+COPY --from=sysroot /opt ${AUTOWARE_SYSROOT}/opt
+COPY --from=sysroot /etc/alternatives ${AUTOWARE_SYSROOT}/etc/alternatives
+
+# Fix Bionic-specific issues
+# ignore pinning package versions for apt
+# hadolint ignore=DL3008
+RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
+    tzdata \
+    gnupg && \
+    rm -rf /var/lib/apt/lists/*
+
+RUN apt-get update && apt-get install -y --no-install-recommends \
+    gcc-aarch64-linux-gnu \
+    g++-aarch64-linux-gnu \
+    default-jre

--- a/tools/Dockerimage.crossbuild
+++ b/tools/Dockerimage.crossbuild
@@ -45,3 +45,6 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
     gcc-aarch64-linux-gnu \
     g++-aarch64-linux-gnu \
     default-jre
+
+RUN find /sysroot/opt/ros/dashing/ -name "*.cmake" -type f -exec sed -i -e "s#/usr/lib#/sysroot/usr/lib#g" {} \;
+RUN find /sysroot/opt/ros/dashing/ -name "*.cmake" -type f -exec sed -i -e "s#/usr/include#/sysroot/usr/include#g" {} \;

--- a/tools/toolchain.cmake
+++ b/tools/toolchain.cmake
@@ -1,0 +1,6 @@
+set(CMAKE_C_COMPILER /usr/bin/aarch64-linux-gnu-gcc)
+set(CMAKE_CXX_COMPILER /usr/bin/aarch64-linux-gnu-g++)
+set(CMAKE_SYSROOT /sysroot/)
+set(TINYXML2_LIBRARY /foo)
+
+# colcon build --merge-install --cmake-args -DCMAKE_TOOLCHAIN_FILE=`pwd`/tools/toolchain.cmake -DBUILD_TESTING=OFF

--- a/tools/toolchain.cmake
+++ b/tools/toolchain.cmake
@@ -1,6 +1,9 @@
 set(CMAKE_C_COMPILER /usr/bin/aarch64-linux-gnu-gcc)
 set(CMAKE_CXX_COMPILER /usr/bin/aarch64-linux-gnu-g++)
 set(CMAKE_SYSROOT /sysroot/)
-set(TINYXML2_LIBRARY /foo)
 
-# colcon build --merge-install --cmake-args -DCMAKE_TOOLCHAIN_FILE=`pwd`/tools/toolchain.cmake -DBUILD_TESTING=OFF
+set(CMAKE_FIND_ROOT_PATH /sysroot)
+set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
+set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE ONLY)
+set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
+set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)

